File: 2020/07_HandyHaversacks/rule.lua
-- ======================================================================
-- Handy Haversacks
--   Advent of Code 2020 Day 07 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                         r u l e . l u a
-- ======================================================================
-- Rule for the Advent of Code 2020 Day 07 puzzle

-- ----------------------------------------------------------------------
--                                                                  local
-- ----------------------------------------------------------------------
local Rule = { part2=false, text='', bag='', bags={} }
-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------
local MATCH_BAG = "%a+ %a+"
local MATCH_REST = "%d+ %a+ %a+"
local MATCH_NUMBER = "%d+"

-- ======================================================================
--                                                                   Rule
-- ======================================================================


function Rule:Rule (o)
  -- Object for Handy Haversacks


  -- 1. Set the initial values
  o = o or {}
  o.part2 = o.part2 or false
  o.text = o.text or {}
  o.bag = ''
  o.bags = {}

  -- 2. Create the object metatable
  self.__index = self
  setmetatable(o, self)

  -- 3. Process text (if any)
  if o.text ~= nil and #o.text > 0 then
     o:_process_text(o.text)
  end

  return o
end

function Rule:_process_text(text)
  -- Assign values from text

  -- 0. Precondition axioms
  assert(text ~= nil and #text > 0)
  
  -- 1. Get the bag color
  self.bag = text:match(MATCH_BAG)
  
  -- 2. Loop for all of the contained bags
  for contained in text:gmatch(MATCH_REST) do
    
    -- 3. Get number abd color of contained bag
    local number = tonumber(contained:match(MATCH_NUMBER))
    local color = contained:match(MATCH_BAG)
    
    -- 4. Save the contained bag
    self.bags[color] = number
  end
  
end

function Rule:contains(color)
  -- Returns the number of the bags that this bag contains
  
  -- 1. Loop for all the contained bags
  for bcolor, bnumber in pairs(self.bags) do
    
    -- 2. If the colors match, return the number
    if color == bcolor then
      return bnumber
    end
  end
   
  -- 3. Not found, return 0
  return 0
end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
return Rule

-- ======================================================================
-- end                         r u l e . l u a                        end
-- ======================================================================

File: 2020/07_HandyHaversacks/rule.ts
// ======================================================================
// Handy Haversacks
//   Advent of Code 2020 Day 07 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           r u l e . t s
//
// Rule for the Advent of Code 2020 Day 07 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------

// ----------------------------------------------------------------------
//                                                                  types
// ----------------------------------------------------------------------
type BagRecord = Record<string, number>;

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
const RE_MATCH = /^([a-z ]+) bags contain ([0-9a-z, ]+).$/;
const RE_BAG = /([0-9]+) ([a-z ]+) bag/g;

// ======================================================================
//                                                                   Rule
// ======================================================================

export class Rule {
  // Object for Handy Haversacks
  text: string;

  part2: boolean;

  bag: string;

  bags: BagRecord;

  constructor(text: string, part2 = false) {
    // Create a Rule object

    // 1. Set the initial values
    this.text = text === undefined ? '' : text;
    this.part2 = part2 === undefined ? false : part2;
    this.bag = '';
    this.bags = {};

    // 2. Process text (if any)
    if (this.text.length !== 0) {
      this.processText();
    }
  }

  processText() {
    // Use regular expressions to split up the input text

    // 1. Break up the text into the parts before and after "contains"
    const matched = this.text.match(RE_MATCH);
    if (!matched) {
      console.log(`Unable to parse: ${this.text}`);
      return;
    }
    // eslint-disable-next-line prefer-destructuring
    this.bag = matched[1];
    const bags = matched[2];

    // 2. Loop for the containing bags
    const matches = Array.from(bags.matchAll(RE_BAG));
    for (let match = 0; match < matches.length; match += 1) {
      const bagNumber = matches[match][1];
      const bagColor = matches[match][2];

      // 3. Save the interier bag
      this.bags[bagColor] = parseInt(bagNumber, 10);
    }
  }

  contains(color: string): number {
    // Return the number of bags of that color that the bag contains
    if (this.bags[color]) {
      return this.bags[color];
    }
    return 0;
  }

  inside(): number {
    // Return the number of bags inside this bag
    // 1. Start with nothing
    let result = 0;
    // 2. Loop for all of the bags
    const values = Object.values(this.bags);
    for (let indx = 0; indx < values.length; indx += 1) {
      // 3. Add in the number of bags
      result += values[indx];
    }
    // 4. Return the number of bags inside this bag
    return result;
  }
}

// ======================================================================
// end                          r u l e . t s                         end
// ======================================================================
